lib/Doctrine/Annotations/DocParser.php

```diff
@@ -468,97 +396,106 @@ private function collectAnnotationMetadata($name)
             ]);
         }
 
-        $class      = new \ReflectionClass($name);
-        $docComment = $class->getDocComment();
-
-        // Sets default values for annotation metadata
-        $metadata = [
-            'default_property' => null,
-            'has_constructor'  => (null !== $constructor = $class->getConstructor()) && $constructor->getNumberOfParameters() > 0,
-            'properties'       => [],
-            'property_types'   => [],
-            'attribute_types'  => [],
-            'targets_literal'  => null,
-            'targets'          => Target::TARGET_ALL,
-            'is_annotation'    => false !== strpos($docComment, '@Annotation'),
-        ];
+        $class          = new \ReflectionClass($name);
+        $docComment     = $class->getDocComment();
+        $constructor    = $class->getConstructor();
+        $useConstructor = $constructor !== null && $constructor->getNumberOfParameters() > 0;
 
         // verify that the class is really meant to be an annotation
-        if ($metadata['is_annotation']) {
-            self::$metadataParser->setTarget(Target::TARGET_CLASS);
+        if (strpos($docComment, '@Annotation') === false) {
+            $this->nonAnnotationClasses[$name] = true;
+            return;
+        }
 
-            foreach (self::$metadataParser->parse($docComment, 'class @' . $name) as $annotation) {
-                if ($annotation instanceof Target) {
-                    $metadata['targets']         = $annotation->targets;
-                    $metadata['targets_literal'] = $annotation->literal;
+        $annotationBuilder = new AnnotationMetadataBuilder($name);
 
-                    continue;
-                }
+        if ($useConstructor) {
+            $annotationBuilder->withUsingConstructor();
```

Comment:
In general I prefer to avoid public properties until PHP adds proper accessors or at least readonly properties. For `@internal` stuff it may be sufficient but since we don't have PHPStan at this moment, I'd leave this to a later stage - once we have PHPStan L7, we can do this -> create a blocker issue for 2.0 for this?